### PR TITLE
Check that an appservice protocol has at least one service providing it

### LINKED ISSUE #9694: /thirdparty/protocols API returned an empty object for an appservice that wasn't ready
### Description
`/thirdparty/protocols` returned 
```
{ irc: {} }
```

The `irc` object is missing a few required values.

https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-thirdparty-protocols

### Steps to reproduce

- I set up my IRC bridge but the pingback port wasn't working yet (I had set it to the wrong port in the appservice file) and the IRC server on localhost wasn't connecting yet.

Synapse-config, Bridge config and logs (look around the time 2021-03-25 20:44):
https://gateway.pinata.cloud/ipfs/QmTbjP9PwfaTELkYQQ1HmznR4n8GNusveE6BwpZXw5tjk4

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from _homeserver.log_, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.
-->

### Version information

<!--
 What version of Synapse is running?

You can find the Synapse version with this command:

$ curl http://localhost:8008/_synapse/admin/v1/server_version

(You may need to replace `localhost:8008` if Synapse is not configured to
listen on that port.)
-->
- **Version**:
- Synapse 1.29.0 with Python 3.8.8
- https://registry.hub.docker.com/layers/matrixdotorg/synapse/latest/images/sha256-67a507a29e0406a0e93a9040617380f7310f625e86adb5652ef24cd1ce39db48

- **Install method**:
- DockerHub's `matrixdotorg/synapse:latest`

- **Platform**:
- Manjaro Linux (based on the repo of Arch Linux)

### Additional notes:

The API response caused an issue in Element Web because it's not spec-compliant.

https://github.com/vector-im/element-web/issues/16789

https://user-images.githubusercontent.com/10872136/112535146-3a7b9100-8dac-11eb-9270-a7f4746d6300.mp4

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes #9694 

If there is no information to serve for a given protocol, we should prefer to omit that protocol rather than returning an empty object. The spec https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-thirdparty-protocols mandates that protocols must contain at least a few fields to be valid.

### CODE REVIEW

File: synapse/handlers/appservice.py [hunk 1]
```diff
@@ -406,7 +406,11 @@ def _merge_instances(infos: List[JsonDict]) -> JsonDict:
 
             return combined
 
-        return {p: _merge_instances(protocols[p]) for p in protocols.keys()}
+        return {
+            p: _merge_instances(protocols[p])
+            for p in protocols.keys()
+            if len(protocols[p]) > 0
```
Comment (clokep) [inline synapse/handlers/appservice.py:412]: This can probably just be:

```suggestion
            if protocols[p]
```

But something else seems bogus here -- `_merge_instances` will return an empty dict if there's no protocols passed in. Should that throw an exception or something if we're explicitly filtering those out?

Comment (Half-Shot) [reply synapse/handlers/appservice.py:412]: I don't think we want the request to fail so throwing seems off. Returning None will return `{"my-protocol": None}` which isn't the same as `{}`, which is what the spec expects us to return if the protocol is incomplete (i.e. it has no appservices implementing it).

Removing the `len` seems sane however.

Comment (clokep) [reply synapse/handlers/appservice.py]: Right, my point was mostly that it was redundant Removing it (like you did) seems reasonable.

File: synapse/handlers/appservice.py [hunk 2]
```diff
@@ -406,7 +403,11 @@ def _merge_instances(infos: List[JsonDict]) -> JsonDict:
 
             return combined
 
-        return {p: _merge_instances(protocols[p]) for p in protocols.keys()}
+        return {
+            p: _merge_instances(protocols[p])
+            for p in protocols.keys()
+            if protocols[p]
```
Comment (clokep) [inline synapse/handlers/appservice.py:409]:
```suggestion
            p: _merge_instances(protocols[p]) for p in protocols.keys() if protocols[p]
```

File: synapse/handlers/appservice.py (hunk 1 above)
Comment (Half-Shot) [reply synapse/handlers/appservice.py]: Ah, I see. Good good.

**Comment by clokep** [comment]
Are there any tests for this code (in sytest or unit tests)? It would be beneficial to add some for edge-cases like this.

**Comment by Half-Shot** [comment]
Complement looked too scary / time consuming for trying to mock out appservice responses, so I've settled for some classic unit tests.

File: changelog.d/10532.bugfix [hunk 3]
```diff
@@ -0,0 +1 @@
+Only report protocols that are implemented by at least one service, as required by the spec when handling `GET /_matrix/client/r0/thirdparty/protocols`.
```
Comment (clokep) [inline changelog.d/10532.bugfix:1]: I assume this has always been broken?

```suggestion
Fix a long-standing bug where protocols which are not implemented by any appservices were incorrectly returned via `GET /_matrix/client/r0/thirdparty/protocols`.
```

**Review by clokep — APPROVED** [review]
LGTM! Thanks for adding tests!

File: changelog.d/10532.bugfix (hunk 3 above)
Comment (Half-Shot) [reply changelog.d/10532.bugfix:1]: Well put :+1:
